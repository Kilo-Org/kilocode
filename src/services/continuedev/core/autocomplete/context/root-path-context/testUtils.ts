import fs from "fs"
import path from "path"
import { fileURLToPath } from "node:url"
import { expect, vi } from "vitest"

const __dirname = path.dirname(fileURLToPath(import.meta.url))

import Parser from "web-tree-sitter"
import type { Position } from "../../../index.js"
import { testIde } from "../../../test/fixtures"
import { getAst, getTreePathAtCursor } from "../../util/ast"
import { ImportDefinitionsService } from "../ImportDefinitionsService"
import { RootPathContextService } from "./RootPathContextService"

function splitTextAtPosition(fileContent: string, position: Position): [string, string] {
	const lines = fileContent.split("\n")
	let currentPos = 0

	// Calculate position based on the provided line and character
	for (let i = 0; i < position.line; i++) {
		currentPos += lines[i].length + 1 // +1 for the newline character
	}
	const splitPos = currentPos + position.character

	return [fileContent.slice(0, splitPos), fileContent.slice(splitPos)]
}

export async function testRootPathContext(
	folderName: string,
	relativeFilepath: string,
	position: Position,
	expectedDefinitionPositions: Parser.Point[],
) {
	// Create a mocked instance of RootPathContextService
	const ide = testIde
	const importDefinitionsService = new ImportDefinitionsService(ide)
	const service = new RootPathContextService(importDefinitionsService, ide)

	const getSnippetsMock = vi
		// @ts-expect-error -- spying on private method for test verification
		.spyOn(service, "getSnippets")
		// @ts-expect-error -- mocking private method implementation
		.mockImplementation((_filePath, _endPosition) => {
			return []
		})

	// Copy the folder to the test directory
	const folderPath = path.join(__dirname, "__fixtures__", folderName)
	let workspaceDir = (await ide.getWorkspaceDirs())[0] as string
	if (workspaceDir.startsWith("file:")) {
		workspaceDir = fileURLToPath(workspaceDir)
	}
	const testFolderPath = path.join(workspaceDir, folderName)
	fs.cpSync(folderPath, testFolderPath, {
		recursive: true,
		force: true,
	})

	// Get results of root path context
	const startPath = relativeFilepath.startsWith("file:")
		? fileURLToPath(relativeFilepath)
		: path.join(testFolderPath, relativeFilepath)
	const [prefix, suffix] = splitTextAtPosition(fs.readFileSync(startPath, "utf8"), position)
	const fileContents = prefix + suffix
	const ast = await getAst(startPath, fileContents)
	if (!ast) {
		throw new Error("AST is undefined")
	}

	const treePath = await getTreePathAtCursor(ast, prefix.length)
	await service.getContextForPath(startPath, treePath)

	expect(getSnippetsMock).toHaveBeenCalledTimes(expectedDefinitionPositions.length)

	expectedDefinitionPositions.forEach((position, index) => {
		expect(getSnippetsMock).toHaveBeenNthCalledWith(
			index + 1,
			expect.any(String), // filepath argument
			position,
			expect.any(String), // language argument
		)
	})
}
