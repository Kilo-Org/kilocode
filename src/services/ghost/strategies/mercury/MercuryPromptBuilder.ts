import * as vscode from "vscode"
import { EditableRegion, EditableRegionResult } from "./EditableRegionCalculator"

/**
 * Builds Mercury prompts with proper code-to-edit markers.
 * This shows exactly what gets sent to Mercury for editing.
 */
export class MercuryPromptBuilder {
	private static readonly MERCURY_MARKERS = {
		OPEN: "<|code_to_edit|>",
		CLOSE: "<|/code_to_edit|>",
		CURSOR: "<|cursor|>",
	}

	/**
	 * Build the current file content block with editable region markers.
	 * This is the core of what Mercury sees - full file with marked editable region.
	 */
	public buildCurrentFileContentBlock(
		document: vscode.TextDocument,
		cursorRange: vscode.Range,
		editableRegion: EditableRegion,
	): string {
		const lines = document.getText().split("\n")
		const cursorLine = cursorRange.start.line
		const cursorCharacter = cursorRange.start.character

		const { OPEN, CLOSE, CURSOR } = MercuryPromptBuilder.MERCURY_MARKERS
		const { editableStart, editableEnd } = editableRegion

		// Build numbered lines with markers
		// Use #| format (matching Continue's approach) for line numbers
		const numberedLines = lines.map((line, index) => {
			let numberedLine = `${index + 1} #| ${line}`

			// Add cursor marker if this is the cursor line
			if (index === cursorLine) {
				const beforeCursor = line.substring(0, cursorCharacter)
				const afterCursor = line.substring(cursorCharacter)
				const lineWithCursor = `${beforeCursor}${CURSOR}${afterCursor}`
				numberedLine = `${index + 1} #| ${lineWithCursor}`
			}

			// Add editable region markers
			if (index === editableStart) {
				numberedLine = `${OPEN}\n${numberedLine}`
			}
			if (index === editableEnd) {
				numberedLine = `${numberedLine}\n${CLOSE}`
			}

			return numberedLine
		})

		return numberedLines.join("\n")
	}

	/**
	 * Build recently viewed code snippets block
	 */
	public buildRecentlyViewedCodeSnippetsBlock(
		recentlyViewedSnippets: Array<{ content: string; filepath: string }>,
	): string {
		if (recentlyViewedSnippets.length === 0) {
			return ""
		}

		const snippetBlocks = recentlyViewedSnippets.map((snippet) => {
			return `File: ${snippet.filepath}\n\`\`\`\n${snippet.content}\n\`\`\``
		})

		return snippetBlocks.join("\n\n")
	}

	/**
	 * Build edit history block
	 */
	public buildEditHistoryBlock(editHistory: string[]): string {
		return editHistory.length > 0 ? editHistory.join("\n\n") : "No recent edit history available."
	}

	/**
	 * Build complete Mercury user prompt
	 */
	public buildUserPrompt(
		document: vscode.TextDocument,
		cursorRange: vscode.Range,
		editableRegion: EditableRegion,
		recentlyViewedSnippets: Array<{ content: string; filepath: string }> = [],
		editHistory: string[] = [],
	): string {
		const recentlyViewedBlock = this.buildRecentlyViewedCodeSnippetsBlock(recentlyViewedSnippets)
		const currentFileBlock = this.buildCurrentFileContentBlock(document, cursorRange, editableRegion)
		const editHistoryBlock = this.buildEditHistoryBlock(editHistory)

		return `<|recently_viewed_code_snippets|>
${recentlyViewedBlock}
<|/recently_viewed_code_snippets|>

<|current_file_content|>
${currentFileBlock}
<|/current_file_content|>

<|edit_diff_history|>
${editHistoryBlock}
<|/edit_diff_history|>`
	}

	/**
	 * Extract just the editable region content that will be compared with Mercury's response
	 */
	public extractEditableRegionForComparison(document: vscode.TextDocument, editableRegion: EditableRegion): string {
		const lines = document.getText().split("\n")
		return lines.slice(editableRegion.editableStart, editableRegion.editableEnd + 1).join("\n")
	}

	/**
	 * Get debug info about prompt construction
	 */
	public getPromptDebugInfo(
		document: vscode.TextDocument,
		editableRegion: EditableRegion,
		recentlyViewedSnippets: Array<{ content: string; filepath: string }> = [],
		editHistory: string[] = [],
	): {
		totalLines: number
		editableLines: number
		editableRange: string
		recentSnippetsCount: number
		editHistoryCount: number
		estimatedTokens: number
	} {
		const totalLines = document.lineCount
		const editableLines = editableRegion.editableEnd - editableRegion.editableStart + 1
		const editableRange = `${editableRegion.editableStart + 1}-${editableRegion.editableEnd + 1}`

		// Rough token estimate (4 chars = 1 token)
		const promptContent = this.buildUserPrompt(
			document,
			new vscode.Range(0, 0, 0, 0),
			editableRegion,
			recentlyViewedSnippets,
			editHistory,
		)
		const estimatedTokens = Math.ceil(promptContent.length / 4)

		return {
			totalLines,
			editableLines,
			editableRange,
			recentSnippetsCount: recentlyViewedSnippets.length,
			editHistoryCount: editHistory.length,
			estimatedTokens,
		}
	}
}
