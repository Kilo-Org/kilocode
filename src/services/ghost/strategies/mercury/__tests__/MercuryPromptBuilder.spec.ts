import { describe, it, expect } from "vitest"
import { MercuryPromptBuilder } from "../MercuryPromptBuilder"
import { parseContentWithCursor } from "./testUtils"

describe("MercuryPromptBuilder", () => {
	let builder: MercuryPromptBuilder

	beforeEach(() => {
		builder = new MercuryPromptBuilder()
	})

	describe("buildCurrentFileContentBlock", () => {
		it("should build file content with editable region markers and cursor", () => {
			const contentWithCursor = `class User {
    constructor(name, age) {
        this.name = name;
        this.age = age;
    }

    getDisplayName() {
        ␣
    }

    isAdult() {
        return this.age >= 18;
    }
}`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const editableRegion = { editableStart: 6, editableEnd: 8, totalLines: 13, tokensUsed: 100 }

			const fileBlock = builder.buildCurrentFileContentBlock(document, cursorRange, editableRegion)

			expect(fileBlock).toMatchSnapshot()
		})

		it("should handle cursor at beginning of file", () => {
			const contentWithCursor = `␣class User {
    constructor() {}
}`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const editableRegion = { editableStart: 0, editableEnd: 2, totalLines: 3, tokensUsed: 50 }

			const fileBlock = builder.buildCurrentFileContentBlock(document, cursorRange, editableRegion)

			expect(fileBlock).toMatchSnapshot()
		})
	})

	describe("buildUserPrompt", () => {
		it("should build complete Mercury prompt with all sections", () => {
			const contentWithCursor = `function test() {
    retu␣rn 42;
}`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const editableRegion = { editableStart: 0, editableEnd: 2, totalLines: 3, tokensUsed: 30 }

			const recentlyViewedSnippets = [{ content: "const x = 1;", filepath: "helper.js" }]
			const editHistory = ["Added function test()"]

			const prompt = builder.buildUserPrompt(
				document,
				cursorRange,
				editableRegion,
				recentlyViewedSnippets,
				editHistory,
			)

			expect(prompt).toMatchSnapshot()
		})

		it("should extract editable region for comparison with Mercury response", () => {
			const { document } = parseContentWithCursor(`line 0
line 1
line ␣2
line 3
line 4`)

			const editableRegion = { editableStart: 1, editableEnd: 3, totalLines: 5, tokensUsed: 50 }

			const extracted = builder.extractEditableRegionForComparison(document, editableRegion)

			expect(extracted).toMatchSnapshot()
		})
	})

	describe("buildRecentlyViewedCodeSnippetsBlock", () => {
		it("should format snippets with file paths", () => {
			const snippets = [
				{ content: "const helper = () => 42;", filepath: "/utils/helper.js" },
				{ content: "interface User {\n  name: string;\n}", filepath: "/types/user.ts" },
			]

			const block = builder.buildRecentlyViewedCodeSnippetsBlock(snippets)

			expect(block).toMatchSnapshot()
		})

		it("should handle empty snippets", () => {
			const block = builder.buildRecentlyViewedCodeSnippetsBlock([])

			expect(block).toMatchSnapshot()
		})
	})
})
