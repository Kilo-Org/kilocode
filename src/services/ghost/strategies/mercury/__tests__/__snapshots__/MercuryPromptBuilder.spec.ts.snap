// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MercuryPromptBuilder > buildCurrentFileContentBlock > should build file content with editable region markers and cursor 1`] = `
"1 #| class User {
2 #|     constructor(name, age) {
3 #|         this.name = name;
4 #|         this.age = age;
5 #|     }
6 #| 
<|code_to_edit|>
7 #|     getDisplayName() {
8 #|         <|cursor|>
9 #|     }
<|/code_to_edit|>
10 #| 
11 #|     isAdult() {
12 #|         return this.age >= 18;
13 #|     }
14 #| }"
`;

exports[`MercuryPromptBuilder > buildCurrentFileContentBlock > should handle cursor at beginning of file 1`] = `
"<|code_to_edit|>
1 #| <|cursor|>class User {
2 #|     constructor() {}
3 #| }
<|/code_to_edit|>"
`;

exports[`MercuryPromptBuilder > buildRecentlyViewedCodeSnippetsBlock > should format snippets with file paths 1`] = `
"File: /utils/helper.js
\`\`\`
const helper = () => 42;
\`\`\`

File: /types/user.ts
\`\`\`
interface User {
  name: string;
}
\`\`\`"
`;

exports[`MercuryPromptBuilder > buildRecentlyViewedCodeSnippetsBlock > should handle empty snippets 1`] = `""`;

exports[`MercuryPromptBuilder > buildUserPrompt > should build complete Mercury prompt with all sections 1`] = `
"<|recently_viewed_code_snippets|>
File: helper.js
\`\`\`
const x = 1;
\`\`\`
<|/recently_viewed_code_snippets|>

<|current_file_content|>
<|code_to_edit|>
1 #| function test() {
2 #|     retu<|cursor|>rn 42;
3 #| }
<|/code_to_edit|>
<|/current_file_content|>

<|edit_diff_history|>
Added function test()
<|/edit_diff_history|>"
`;

exports[`MercuryPromptBuilder > buildUserPrompt > should extract editable region for comparison with Mercury response 1`] = `
"line 1
line 2
line 3"
`;
